op-batcher/channel_manager.go

```diff
@@ -35,90 +69,174 @@ func (s *channelManager) Clear() {
+		// We need to keep track of stale transactions instead
+		return
+	}
+	delete(s.pendingTransactions, id)
+	s.confirmedTransactions[id] = inclusionBlock
+
+	// If this channel timed out, put the pending blocks back into the local saved blocks
+	// and then reset this state so it can try to build a new channel.
+	if s.pendingChannelIsTimedOut() {
+		s.log.Warn("Channel timed out", "id", s.pendingChannel)
+		s.blocks = append(s.pendingBlocks, s.blocks...)
+		s.clearPendingChannel()
+	}
+	// If we are done with this channel, record that.
+	if s.pendingChannelIsFullySubmitted() {
+		s.log.Info("Channel is fully submitted", "id", s.pendingChannel)
+		s.clearPendingChannel()
+	}
+}
 
-// TxData returns the next tx.data that should be submitted to L1.
-// It is very simple & currently ignores the l1Head provided (this will change).
-// It may buffer very large channels as well.
-func (s *channelManager) TxData(l1Head eth.L1BlockRef) ([]byte, txID, error) {
-	// Note: l1Head is not actually used in this function.
+// TODO: Create separate "pending" state
+func (s *channelManager) clearPendingChannel() {
+	s.channelPending = false
+	s.pendingChannel = derive.ChannelID{}
+	s.pendingBlocks = nil
+	s.pendingTransactions = make(map[txID][]byte)
+	s.confirmedTransactions = make(map[txID]eth.BlockID)
+}
 
-	// Return a pre-existing frame if we have it.
-	if len(s.datas) != 0 {
-		r := s.datas[0]
-		s.datas = s.datas[1:]
-		return r.data, r.id, nil
+func (s *channelManager) pendingChannelIsTimedOut() bool {
+	if !s.channelPending {
+		return false // no channel to be timed out
+	}
+	min := uint64(math.MaxUint64)
+	max := uint64(0)
+	for _, inclusionBlock := range s.confirmedTransactions {
+		if inclusionBlock.Number < min {
+			min = inclusionBlock.Number
+		}
+		if inclusionBlock.Number > max {
+			max = inclusionBlock.Number
+		}
 	}
+	return max-min >= s.channelTimeout
+}
 
-	// Also return io.EOF if we cannot create a channel
-	if len(s.blocks) == 0 {
-		return nil, txID{}, io.EOF
+func (s *channelManager) pendingChannelIsFullySubmitted() bool {
+	if !s.channelPending {
+		return false // todo: can decide either way here. Nonsensical answer though
 	}
+	return len(s.pendingTransactions)+len(s.datas) == 0
+}
+
+func (s *channelManager) TxFailed(id txID) {
+	if data, ok := s.pendingTransactions[id]; ok {
+		s.datas = append(s.datas, taggedData{data, id})
+		delete(s.pendingTransactions, id)
+	} else {
+		s.log.Warn("marked transaction as failed despite having no record of it.")
+	}
+}
 
-	// Add all pending blocks to a channel
+// blocksToFrames turns a set of blocks into a set of frames inside a channel.
+// It will only create a single channel which contains up to `MAX_RLP_BYTES`. Any
+// blocks not added to the channel are returned. It uses the max supplied frame size.
+func blocksToFrames(blocks []*types.Block, maxFrameSize uint64) (derive.ChannelID, [][]byte, []*types.Block, error) {
```

Comment:
Why immediately produce the frames, instead of holding on to the `derive.ChannelOut`? You can produce frames with the exact size we want later that way, and pack more efficiently?